docs/cloud-docker-images.md

```diff
@@ -0,0 +1,100 @@
+overhead of waiting for the build & push.
+- `//my_repo/my_package:my_target_image`: builds the image, but doesn't push it
+or run your script
+- `//my_repo/my_package:my_target_push`: builds and pushes the image, but
+doesn't run your script
+
+#### Custom base images
+The `sematic_pipeline` macro also allows you to specify a custom base image to
+cover any dependencies you have that aren't specified in bazel. You can do this
+by setting the `base` field of the `sematic_pipeline` macro, If, for
+example, your bazel setup assumes that it is running on a machine with a
+particular version of Cuda, you can bake that into the base image but let
+Sematic's bazel hookup handle all the python and other native code you want in
+the image. Sematic also provides a couple base images out-of-the-box, which can
+be referenced as `"@sematic-worker-base//image"` and
+`"@sematic-worker-cuda//image"` in bazel. `sematic-worker-cuda` includes
+python 3.9 and an installation of cuda, and is actually the base image used by
+default when you use the `sematic_pipeline` macro. `sematic-worker-base` is the
+same, but doesn't include cuda (this is helpful if you want a lighter image and
+don't use GPUs in your pipeline).
+
+If you like, you can also specify another base image not provided by Sematic.
+There are a few requirements on this image though, so be sure to read through
+the ["Totally Custom Image"](#totally-custom-image) section below to understand
+what those are. Note that if you are using bazel, requirements 1 & 2 from that
+section will already be taken care of.
+
+### requirements.txt
+Coming soon!
+
+### Totally Custom Image
+If you want full control over how your Docker image is produced, Sematic
+provides a hook to make that possible. Just set the `SEMATIC_CONTAINER_IMAGE`
+environment variable to the URI for the Docker image you want your pipeline to
+use. There are some requirements on this image though:
+
+1. It must contain your source code and its dependencies
+2. It must be pushed to a container registry that can be accessed by the cluster
+where your code is going to run in the cloud
+3. `/usr/bin/python3` must be a valid python interpreter in the image
+4. The home directory inside the image must be writable
+
+## Working with images
+When launching Sematic pipelines, there is always a python script that submits
+the job (either for local exection or execution in the cloud). We refer to this
+as the "launch script." You may want to have your launch script behave
+differently depending on whether a suitable cloud image is available. Sematic
+has provided `sematic.has_cloud_image` to enable this use case. A common pattern
+is to determine which resolver to use based on the result of this function:
+
+```python
+from sematic import has_cloud_image, CloudResolver
+
+from my_package import my_pipeline
+
+resolver = CloudResolver() if has_cloud_image() else None
+my_pipeline().resolve(resolver)
+```
```

Comment:
Should we abstract this even further, with e.g. an `AutoResolver`?

```python
def AutoResolver(**kwargs):
    return CloudResolver(**kwargs) if has_cloud_image() else LocalResolver(**kwargs)
```
(pseudocode)

Reply:
I'm skeptical of choosing a resolver *automatically* based on the presence or absence of an image. I'd rather have a more explicit signal of "this execution should use the cloud resolver." Ideally I would like to wrap launching in the Sematic CLI and give it options you can use to select a resolver from the command line. We can make it more robust in that case. This seems like a fine pattern to me for now. WDYT?